### PR TITLE
Newest Export Lookup is Batched

### PR DESCRIPTION
**Pull request by krishnaglick**
## Change description

The magic of sql

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: core/src/modules/ops/destination.ts [hunk 1]
```diff
@@ -903,15 +902,16 @@ export namespace DestinationOps {
     return { success, error, retryexportIds, retryDelay };
   }
 
-  async function cancelOldExport(oldExport: Export, newExport: Export) {
```
Comment (krishnaglick) [inline core/src/modules/ops/destination.ts:906]: We don't have the exact new export now.

Comment (tealjulia) [reply core/src/modules/ops/destination.ts:906]: Noting that this means the cancel messaging is different now too

File: core/__tests__/models/destination/plugins/exportRecords.ts [hunk 2]
```diff
@@ -738,8 +738,8 @@ describe("models/destination - with custom exportRecords plugin", () => {
         errorLevel: null,
       });
 
-      helper.changeTimestamps([exportOne], true, new Date());
-      helper.changeTimestamps(
+      await helper.changeTimestamps([exportOne], true, new Date());
+      await helper.changeTimestamps(
```
Comment (evantahler) [inline core/__tests__/models/destination/plugins/exportRecords.ts:742]: good catch!

File: core/src/modules/ops/destination.ts [hunk 3]
```diff
@@ -926,23 +926,37 @@ export namespace DestinationOps {
...
+          "id"
+        FROM (
+          SELECT
+            "id",
+            ROW_NUMBER() OVER (PARTITION BY "exports"."recordId" ORDER BY "exports"."createdAt" DESC) AS __rownum
+          FROM
+            "exports"
+          WHERE
+            "exports"."destinationId" = ?
+            AND "exports"."id" IN (?)) AS __ranked
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: I had no idea you could use Sequelize's string helpers for custom queries! That's powerful.... and there are probably some other places to update to do this.

Comment (evantahler) [reply core/src/modules/ops/destination.ts]: 👍 having the where clause on the inner query vs the outer query

Comment (evantahler) [reply core/src/modules/ops/destination.ts]: I don't think this part is quire right
```sql
AND "exports"."id" IN (?)) AS __ranked
```

We want to find the newest export that isn't necessarily in  `givenExports`
```sql
AND "exports"."recordId" IN (?)) AS __ranked
```

So the values would be 

```
values: [destination.id, givenExports.map(({ recordId }) => recordId)],
```

Comment (evantahler) [reply core/src/modules/ops/destination.ts]: The tests should be updated to catch something like the above.  Have a later export that isn't in the batch being considered.

**Review by evantahler — CHANGES_REQUESTED** [review]

File: core/src/modules/ops/destination.ts (hunk 3 above)
Comment (krishnaglick) [reply core/src/modules/ops/destination.ts]: Done!

**Review by teallarson — APPROVED** [review]

**Review by evantahler — COMMENTED** [review]
I think the tests are showing that something isn't quite right yet

**Comment by krishnaglick** [comment]
> I think the tests are showing that something isn't quite right yet

Seemed to be an async issue, sorted out now.

File: core/src/modules/ops/destination.ts [hunk 4]
```diff
@@ -926,23 +926,40 @@ export namespace DestinationOps {
...
+        WHERE
+          "__ranked"."__rownum" = 1;`,
+            values: [
+              destination.id,
+              givenExports.map(({ recordId }) => recordId),
+            ],
+          },
+          undefined
+        )
+        .then((exports) => exports[0].map((e) => e.id));
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: I think you can skip this `[0]` if you add this kind of thing -> https://github.com/grouparoo/grouparoo/blob/main/core/src/modules/ops/record.ts#L1062-L1065 to get an object per row.

I think what you are getting back now is something like  rows -> exports -> id

File: core/src/modules/ops/destination.ts [hunk 5]
```diff
@@ -926,23 +926,40 @@ export namespace DestinationOps {
     const locks: Awaited<ReturnType<typeof getLock>>[] = [];
 
     try {
+      const exportIds = await Export.sequelize
```
Comment (evantahler) [inline core/src/modules/ops/destination.ts]: It's pedantic, but this SQL is kind of complex - a better variable name might be something like `newestExportIds`

File: core/src/modules/ops/destination.ts (hunk 4 above)
Comment (krishnaglick) [reply core/src/modules/ops/destination.ts]: Oh, I like this!

**Review by evantahler — APPROVED** [review]
If the test are 🟢 I'm 🟢 !

**Pull request merged**